# EDA

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
#import geopandas as gpd
#import plotly.express as px
#import calendar
#import descartes
import os, sys

from matplotlib.ticker import FuncFormatter
#from shapely.geometry import Point, Polygon

In [2]:
os.chdir("../..")
os.chdir("src/utils")

In [3]:
import eda

In [4]:
data = pd.read_csv('../../data/Rodent_Inspection.csv')

In [5]:
data.head()

INSPECTION_TYPE  JOB_TICKET_OR_WORK_ORDER_ID     JOB_ID  JOB_PROGRESS  \
0         Initial                     13252142  PC7732347             1   
1      Compliance                     12802045  PC7500960             2   
2      Compliance                     12990129  PC7054124             2   
3         Initial                     12328443  PC6737067             1   
4         Initial                     11362643  PC7466087             1   

          BBL  BORO_CODE  BLOCK  LOT HOUSE_NUMBER         STREET_NAME  \
0  2024090050          2   2409   50         0000      MELROSE AVENUE   
1  5011840160          5   1184  160          000      WENLOCK STREET   
2  3089140016          3   8914   16          000         Abbey Court   
3  2025390132          2   2539  132           10  MAJ WM DEEGAN BLVD   
4  5002210035          5    221   35          000          Roe Street   

   ZIP_CODE    X_COORD   Y_COORD  LATITUDE  LONGITUDE        BOROUGH  \
0       NaN        NaN       NaN       0.0        0.0          Bronx   
1       NaN        NaN       NaN       0.0        0.0  Staten Island   
2       NaN        0.0       0.0       0.0        0.0       Brooklyn   
3       NaN  1003704.0  240136.0       0.0        0.0          Bronx   
4       NaN   950510.0  169320.0       0.0        0.0  Staten Island   

          INSPECTION_DATE              RESULT           APPROVED_DATE  \
0  07/07/2021 10:05:45 AM  Failed for Other R  07/09/2021 11:33:07 AM   
1  11/10/2011 02:50:52 PM  Failed for Other R  11/14/2011 03:28:40 PM   
2  09/26/2017 02:00:25 PM              Passed  09/27/2017 12:58:55 PM   
3  09/12/2017 02:45:42 PM              Passed  09/14/2017 02:41:03 PM   
4  06/02/2016 03:08:18 PM              Passed  06/07/2016 02:35:38 PM   

     LOCATION  
0  (0.0, 0.0)  
1  (0.0, 0.0)  
2         NaN  
3  (0.0, 0.0)  
4  (0.0, 0.0)

In [6]:
len(data)

2007833

**Al día 16 de noviembre de 2021 contamos con 2,007,833 datos**

In [7]:
data.columns

Index([u'INSPECTION_TYPE', u'JOB_TICKET_OR_WORK_ORDER_ID', u'JOB_ID',
       u'JOB_PROGRESS', u'BBL', u'BORO_CODE', u'BLOCK', u'LOT',
       u'HOUSE_NUMBER', u'STREET_NAME', u'ZIP_CODE', u'X_COORD', u'Y_COORD',
       u'LATITUDE', u'LONGITUDE', u'BOROUGH', u'INSPECTION_DATE', u'RESULT',
       u'APPROVED_DATE', u'LOCATION'],
      dtype='object')

Los tipos de dato que tiene cada variable inicialmente son:

In [8]:
data.dtypes

INSPECTION_TYPE                 object
JOB_TICKET_OR_WORK_ORDER_ID      int64
JOB_ID                          object
JOB_PROGRESS                     int64
BBL                              int64
BORO_CODE                        int64
BLOCK                            int64
LOT                              int64
HOUSE_NUMBER                    object
STREET_NAME                     object
ZIP_CODE                       float64
X_COORD                        float64
Y_COORD                        float64
LATITUDE                       float64
LONGITUDE                      float64
BOROUGH                         object
INSPECTION_DATE                 object
RESULT                          object
APPROVED_DATE                   object
LOCATION                        object
dtype: object

Cambiamos el formato de las columnas

def clean_column(col):
    """ Replace column's spaces and slash with underscore"""
    return col.lower()\
              .replace('/','_')\
              .replace(' ','_')

data.rename(columns={col: clean_column(col) for col in data.columns.values}, inplace=True)

## Perfilamiento general con los datos en crudo

In [9]:
eda.genera_profiling_general(data)

El dataframe tiene 20 columnas.
Hay 12 columnas que tienen NA's.


Estadísticas Resultado
0                  Total de variables        20
1             Conteo de observaciones   2007833
2                     Total de celdas  40156660
3     Cantidad de variables numéricas        11
4      Cantidad de variables de fecha         0
5   Cantidad de variables categóricas         0
6      Cantidad de variables de texto         9
7                   Valores faltantes    334484
8     Porcentaje de valores faltantes        0%
9                Renglones duplicados         0
10   Porcentaje de valores duplicados        0%

En la base de datos tenemos 334,484 elementos faltantes, distribuidos de la siguiente forma:

In [10]:
eda.cuenta_nulos_por_columnas(data)

El dataframe tiene 20 columnas.
Hay 12 columnas que tienen NA's.


Missing Values  % del Total
X_COORD                  123209          6.1
Y_COORD                  123209          6.1
HOUSE_NUMBER              48758          2.4
ZIP_CODE                  11154          0.6
LOCATION                   7397          0.4
LATITUDE                   6754          0.3
LONGITUDE                  6754          0.3
STREET_NAME                5483          0.3
APPROVED_DATE              1633          0.1
BOROUGH                     103          0.0
INSPECTION_DATE              17          0.0
RESULT                       13          0.0

Cambiamos tipos de datos

In [11]:
data = data.astype({"INSPECTION_TYPE":'category',
                    "STREET_NAME":'category', 
                    "ZIP_CODE":'category', 
                    "BOROUGH":'category', 
                    "RESULT":'category'})
data['INSPECTION_DATE'] = pd.to_datetime(data['INSPECTION_DATE'], dayfirst=True)
data['APPROVED_DATE'] = pd.to_datetime(data['APPROVED_DATE'], dayfirst=True)

Generamos perfilamiento por tipo de variable

In [12]:
profiling_numericas, profiling_categoricas, profiling_texto, profiling_fechas = \
eda.genera_profiling_por_variable(data)

In [13]:
profiling_numericas

Métrica JOB_TICKET_OR_WORK_ORDER_ID  \
0                             Tipo                       int64   
1          Número de observaciones                 2.00783e+06   
2                            Media                 1.07449e+07   
3              Desviación estándar                 3.52033e+06   
4                      Cuartil 25%                 1.15108e+07   
5                      Cuartil 50%                 1.20439e+07   
6                      Cuartil 75%                 1.25876e+07   
7                           Mínimo                 2.37227e+06   
8                           Máximo                 1.32944e+07   
9   Número de observaciones únicas                     2007833   
10             Número de faltantes                           0   
11                    Top1/veces/%         [12584959, 1, 0.0%]   
12                    Top2/veces/%          [2764783, 1, 0.0%]   
13                    Top3/veces/%          [2809841, 1, 0.0%]   
14                    Top4/veces/%          [2805747, 1, 0.0%]   
15                    Top5/veces/%          [2815988, 1, 0.0%]   

              JOB_PROGRESS                       BBL             BORO_CODE  \
0                    int64                     int64                 int64   
1              2.00783e+06               2.00783e+06           2.00783e+06   
2                     1.92                2.2368e+09                  2.21   
3                     3.71               1.08277e+09                  1.07   
4                        1               1.01922e+09                     1   
5                        1               2.03171e+09                     2   
6                        2                3.0202e+09                     3   
7                        1                     1e+09                     1   
8                      175               9.05361e+09                     9   
9                      175                    296006                     6   
10                       0                         0                     0   
11  [1, 1,445,313, 71.98%]  [1003020001, 438, 0.02%]  [1, 641,019, 31.93%]   
12    [2, 283,901, 14.14%]  [1001650001, 421, 0.02%]  [2, 596,696, 29.72%]   
13      [3, 71,667, 3.57%]  [3001390001, 415, 0.02%]  [3, 531,773, 26.48%]   
14      [4, 57,269, 2.85%]  [1003890036, 383, 0.02%]   [4, 182,533, 9.09%]   
15      [5, 45,782, 2.28%]  [1004200001, 381, 0.02%]    [5, 55,709, 2.77%]   

                   BLOCK                  LOT                 X_COORD  \
0                  int64                int64                 float64   
1            2.00783e+06          2.00783e+06             1.88462e+06   
2                2579.37                  193             1.01993e+06   
3                1996.42              1029.68             1.93368e+06   
4                   1396                   17                  995025   
5                   2176                   36             1.00301e+06   
6                   3219                   60             1.01229e+06   
7                      0                    0                       0   
8                  27000                 9999             3.19139e+08   
9                  12368                 1253                   83782   
10                     0                    0                  123209   
11  [2877, 3,388, 0.17%]   [1, 97,715, 4.87%]     [0.0, 1,154, 0.06%]   
12  [2879, 3,293, 0.16%]  [29, 29,974, 1.49%]  [986072.0, 429, 0.02%]   
13  [2782, 3,241, 0.16%]  [10, 29,137, 1.45%]  [997892.0, 316, 0.02%]   
14  [3343, 2,883, 0.14%]   [5, 29,073, 1.45%]  [986393.0, 310, 0.02%]   
15  [3300, 2,855, 0.14%]  [11, 28,658, 1.43%]  [982706.0, 262, 0.01%]   

                   Y_COORD                          LATITUDE  \
0                  float64                           float64   
1              1.88462e+06                       2.00108e+06   
2                   233848                             40.54   
3              1.89098e+06                              2.98   
4              

In [14]:
profiling_categoricas

Métrica                                   INSPECTION_TYPE  \
0                     Tipo                                          category   
1     Número de categorías                                                 5   
2  Número de observaciones                                           2007833   
3      Observaciones nulas                                                 0   
4    % Observaciones nulas                                                 0   
5           Valores únicos  [Initial, Compliance, BAIT, STOPPAGE, CLEAN_UPS]   
6            Moda1/veces/%                        [Initial, 307,794, 15.33%]   
7            Moda2/veces/%                              [BAIT, 1,585, 0.08%]   
8            Moda3/veces/%                     [Compliance, 280,931, 13.99%]   

                                         STREET_NAME  \
0                                           category   
1                                              12082   
2                                            2007833   
3                                               5483   
4                                                  0   
5  [MELROSE AVENUE, WENLOCK STREET, Abbey Court, ...   
6                             [BROADWAY, 741, 0.04%]   
7                                [3 AVENUE, 4, 0.0%]   
8                         [GRAND CONCOURSE, 1, 0.0%]   

                                            ZIP_CODE  \
0                                           category   
1                                                229   
2                                            2007833   
3                                              11154   
4                                                  0   
5  [nan, 10025.0, 11215.0, 10009.0, 10301.0, 1030...   
6                            [10457.0, 3,864, 0.19%]   
7                                 [11221.0, 1, 0.0%]   
8                                 [10458.0, 1, 0.0%]   

                                             BOROUGH  \
0                                           category   
1                                                  5   
2                                            2007833   
3                                                103   
4                                                  0   
5  [Bronx, Staten Island, Brooklyn, Manhattan, Qu...   
6                       [Manhattan, 596,696, 29.72%]   
7                           [Bronx, 531,773, 26.49%]   
8                        [Brooklyn, 641,019, 31.93%]   

                                              RESULT  
0                                           category  
1                                                  7  
2                                            2007833  
3                                                 13  
4                                                  0  
5  [Failed for Other R, Passed, Rat Activity, Bai...  
6                          [Passed, 271,199, 13.51%]  
7                       [Bait applied, 1,585, 0.08%]  
8                     [Rat Activity, 197,036, 9.81%]

In [15]:
profiling_texto

Métrica   JOB_ID HOUSE_NUMBER LOCATION
0                     Tipo   object       object   object
1  Número de observaciones  2007833      2007833  2007833
2     Observaciones únicas  1445313        21308   311840
3   % Observaciones únicas        0            0        0
4          Tamaño promedio        9      3.35165  34.6904
5            Tamaño mínimo        9            1       10
6            Tamaño máximo        9           47       38

In [16]:
profiling_fechas

Métrica                  INSPECTION_DATE  \
0                            Tipo                   datetime64[ns]   
1         Número de observaciones                          2007816   
2                          Mínimo              1918-10-19 14:34:44   
3                          Máximo              2029-01-01 09:35:51   
4  Número de observaciones únicas                          1951381   
5             Número de faltantes                               17   
6                    Top1/veces/%  [2011-01-10 12:00:00, 30, 0.0%]   
7                    Top2/veces/%  [2011-11-19 12:00:00, 28, 0.0%]   
8                    Top3/veces/%  [2011-12-11 12:00:00, 28, 0.0%]   

                     APPROVED_DATE  
0                   datetime64[ns]  
1                          2006200  
2              2009-01-10 08:06:55  
3              2021-12-11 17:14:04  
4                          1077711  
5                             1633  
6  [2010-08-13 16:09:45, 61, 0.0%]  
7  [2010-08-17 14:31:44, 52, 0.0%]  
8  [2010-08-27 15:08:45, 50, 0.0%]

## Perfilamiento general haciendo limpieza de datos